### PR TITLE
Add nose support for TestCase.addClassCleanup

### PR DESCRIPTION
**Pull request by millerdev**
Since Python 3.8, `TestCase` has a new [`addClassCleanup(...)` method](https://docs.python.org/3.9/library/unittest.html#unittest.TestCase.addClassCleanup), which can be used in `setUpClass` to register cleanup functions. Unfortunately nosetests overrides the part of Python's test suite runner that normally calls `TestCase.doClassCleanups()`. This adds a plugin that calls it at the proper times.

FYI @dimagi/team-commcare-hq 

### Automated test coverage

Yes.

### QA Plan

No QA.

### Rollback instructions

Tests that use `addClassCleanup` should be updated to not use that since it will not be called if this is reverted.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by dannyroberts — APPROVED** [review]

File: corehq/tests/test_classcleanup.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    def decorator(func):
+        @wraps(func)
+        def wrapper(self):
+            func(self)
+            test.result.append(func.__name__)
+            if func.__name__ in test.errors:
+                test.result[-1] += f" {test.error_class.__name__}"
+                raise test.error_class
+        return wrapper
+    return decorator
```
Comment (mjriley) [inline corehq/tests/test_classcleanup.py:142]: I found this decorator a bit confusing, as it's doing two things at once -- it's responsible for logging execution, as well as for raising an error, but it's name reflects only the latter responsibility (and it seems like its primary responsibility is the former). Not a blocker, but I could see this being two decorators -- a logger, which logs the path and whether or not it received an exception, and a raiser, which triggers the exception itself. Just reading over `makeSuite`, it wasn't intuitive to me why the methods were decorated with `maybeError`

File: corehq/tests/test_classcleanup.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+    def setUp(self):
+        pass  # super().setUp() is called by self.run_with_errors(...)
+
+    def makeSuite(self):
+        class Test(TestCase):
+            @classmethod
+            @maybe_error(self)
+            def setUpClass(cls):
+                cls.addClassCleanup(lambda: self.result.append("cleanup"))
```
Comment (mjriley) [inline corehq/tests/test_classcleanup.py:22]: Nit: given that 'cleanup' already exists on a test-by-test basis, it might be more intuitive if this was called 'classCleanup'

File: corehq/tests/test_classcleanup.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            "setUpClass",
+            "setUp",
+            "runTest Exception",
+            "tearDown",
+            "tearDownClass",
+            "cleanup",
+        ])
+
+    def test_cleanup_on_test_fail(self):
+        self.run_with_errors("runTest", error_class=AssertionError)
```
Comment (mjriley) [inline corehq/tests/test_classcleanup.py:88]: Why are we specifically testing on `AssertionError` here? It doesn't seem relevant to this test.

File: corehq/tests/test_classcleanup.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            @classmethod
+            @maybe_error(self)
+            def tearDownClass(cls):
+                pass
+
+        return [Test()]
+
+    def run_with_errors(self, *errors, error_class=Exception):
+        self.errors = errors
+        self.result = []
```
Comment (mjriley) [inline corehq/tests/test_classcleanup.py:45]: Nit: `result` is fairly generic here. It might be more straight-forward if it was like `executedSteps`, so that the assertion would be `eq(self.executedSteps, ['setUpClass', ...])`

File: corehq/tests/test_classcleanup.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    activate = ''  # Activate option not needed. Plugin is always enabled.
+    plugins = [mod.ClassCleanupPlugin()]
+
+    def setUp(self):
+        pass  # super().setUp() is called by self.run_with_errors(...)
+
+    def makeSuite(self):
+        class Test(TestCase):
+            @classmethod
+            @maybe_error(self)
```
Comment (mjriley) [inline corehq/tests/test_classcleanup.py:20]: I had no idea that `makeSuite` was a thing...and it's apparently [deprecated](https://github.com/python/cpython/blob/b50322d20337ca468f2070eedb051a16ee1eba94/Lib/unittest/loader.py#L488-L495) (and difficult to find documentation on).  Would it make sense, in each test, to construct this custom `Test` class manually and then execute it?

File: corehq/tests/test_classcleanup.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
+from functools import wraps
+from unittest import TestCase
+
+from nose.plugins import PluginTester
+from testil import eq
```
Comment (mjriley) [inline corehq/tests/test_classcleanup.py:5]: what is gained by using `eq` here vs relying on `TestCase`'s assertions? I remember you mentioned boilerplate, but everything is already part of a `TestCase`, and each test has `self` passed in. Is this just a preference for writing `eq(` rather than `self.assertEquals(`?

**Review by mjriley — COMMENTED** [review]
I like these tests (I feel like I've seen logging the steps somewhere before...). My main concern is using the deprecated `makeSuite`. Thankfully, this doesn't generate more warnings, but the developers have been telling people not to use this since [at least 2008](https://bugs.python.org/issue2721)

File: corehq/tests/test_classcleanup.py (hunk 1 above)
Comment (millerdev) [reply corehq/tests/test_classcleanup.py]: https://github.com/dimagi/commcare-hq/pull/31084/commits/fbae216d5c9be5ff0a92871bd1f9493d6f6ee8cd

I renamed it to hopefully make it clearer, but didn't split it into two separate decorators because that would either require more logic to protect the logger from errors or tightly couple the decorators by adding significance the order in which they are applied. Also two decorators felt like superfluous boilerplate.

File: corehq/tests/test_classcleanup.py (hunk 2 above)
Comment (millerdev) [reply corehq/tests/test_classcleanup.py]: https://github.com/dimagi/commcare-hq/pull/31084/commits/f8e57b130b00e1d47bfdda5843fb3ce4a16021ba

File: corehq/tests/test_classcleanup.py (hunk 3 above)
Comment (millerdev) [reply corehq/tests/test_classcleanup.py:99]: Tests have two different failure modes: FAIL (test raised `AssertionError`) and ERROR (test raised any other error). I wanted to make sure the different code paths for FAIL vs ERROR handling did not affect class cleanup.

File: corehq/tests/test_classcleanup.py (hunk 5 above)
Comment (millerdev) [reply corehq/tests/test_classcleanup.py]: I think you're thinking of `unittest.makeSuite`, which this is not. This is overriding [`PluginTester.makeSuite`](https://nose.readthedocs.io/en/latest/plugins/testing.html), which must be implemented if `self.suitepath` is `None`.

File: corehq/tests/test_classcleanup.py (hunk 6 above)
Comment (millerdev) [reply corehq/tests/test_classcleanup.py:5]: I copied the pattern from [another plugin test](https://github.com/dimagi/commcare-hq/blob/f8e57b130b00e1d47bfdda5843fb3ce4a16021ba/corehq/tests/test_nose_timing.py). It's not a strong reason, `self.assertEquals()` would have worked too.

**Review by mjriley — APPROVED** [review]
